velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,497 @@
+      ROW({"c0"}, {ARRAY(SMALLINT())}),
+      lambdaTypes,
+      lambdaTypes);
+  testExpression(
+      "reduce(c0, 15 * 3, (s, x) -> s + x * abs(-2), s -> s)",
+      "45 + cast(array_sum_propagate_element_null(transform(c0, x -> x * 2)) AS BIGINT)",
+      ROW({"c0"}, {ARRAY(TINYINT())}),
+      {TINYINT(), TINYINT()},
+      {TINYINT()});
+}
+
+/// Test to ensure session queryCtx is used during expression optimization.
+TEST_F(ExprOptimizerTest, queryCtx) {
+  setQueryTimeZone("Pacific/Apia");
+  testExpression("hour(from_unixtime(9.98489045321E8))", "3");
+  testExpression("minute(from_unixtime(9.98489045321E8))", "4");
+
+  setQueryTimeZone("America/Los_Angeles");
+  testExpression("hour(from_unixtime(9.98489045321E8))", "7");
+  testExpression("minute(from_unixtime(9.98489045321E8))", "4");
+}
+
+/// Test cast optimization that avoids expression evaluation when input to
+/// cast is same as the type of cast expression.
+TEST_F(ExprOptimizerTest, castOptimization) {
+  auto testCast = [&](const TypePtr& type) {
+    const auto rowType = ROW({"a"}, {type});
+    if (type->isPrimitiveType()) {
+      testExpression(
+          fmt::format("cast(a as {})", type->toString()), "a", rowType);
+    } else {
+      std::vector<core::TypedExprPtr> inputs = {
+          std::make_shared<core::FieldAccessTypedExpr>(type, "a")};
+      const auto castExpr =
+          std::make_shared<core::CastTypedExpr>(type, inputs, false);
+      testExpression(castExpr, inputs.at(0), rowType);
+    }
+  };
+
+  // Primitive types.
+  testCast(TINYINT());
+  testCast(INTEGER());
+  testCast(BIGINT());
+  testCast(VARCHAR());
+  testCast(VARBINARY());
+  testCast(DOUBLE());
+  testCast(DECIMAL(5, 2));
+  testCast(DECIMAL(25, 10));
+  testCast(DATE());
+  testCast(TIMESTAMP_WITH_TIME_ZONE());
+
+  // Complex types.
+  testCast(ARRAY(DATE()));
+  testCast(MAP(DOUBLE(), VARCHAR()));
+  testCast(ROW({BIGINT(), DECIMAL(5, 2), TIME()}));
+}
+
+/// Test to validate failing subexpressions are replaced with fail expression.
+TEST_F(ExprOptimizerTest, failExpression) {
+  // Primitive types.
+  const auto divideByZero = "division by zero";
+  assertFailCall(
+      "0 / 0", "cast(fail(division by zero) as BIGINT)", divideByZero);
+  assertFailCall(
+      "CAST(4000000000 as INTEGER)",
+      "cast(fail(Cannot cast BIGINT '4000000000' to INTEGER. Overflow during arithmetic conversion: ) as INTEGER)",
+      "Cannot cast BIGINT '4000000000' to INTEGER. Overflow during arithmetic conversion: ");
+  assertFailCall(
+      "CAST(-4000000000 as INTEGER)",
+      "cast(fail(Cannot cast BIGINT '-4000000000' to INTEGER. Negative overflow during arithmetic conversion: ) as INTEGER)",
+      "Cannot cast BIGINT '-4000000000' to INTEGER. Negative overflow during arithmetic conversion: ");
+  assertFailCall(
+      "(INTEGER '400000' * INTEGER '200000')",
+      "cast(fail(integer overflow: 400000 * 200000) as INTEGER)",
```

Comment:
Write a lambda to build the cast expression with the fail message so that its easier to read the parameters.